### ISSUE TITLE
Add lower limit and upper limit for time series forecasting

### ISSUE THREAD

**Issue by ngupta23** [pycaret/pycaret#1766]
Time series data can be naturally clamped at a limit - could be lower limit or upper limit. 
Provide an option to the user to set this limit such that the predictions and intervals do not cross it.

Reference [Link](https://www.linkedin.com/feed/update/urn:li:activity:6858135434724159488?commentUrn=urn%3Ali%3Acomment%3A%28activity%3A6858135434724159488%2C6858178529432100864%29&replyUrn=urn%3Ali%3Acomment%3A%28activity%3A6858135434724159488%2C6858189510220541952%29)

**Comment by vvvijaya** [comment]
hello, 
Is the enhancement for setting the max/min limit on the data itself ("Time series data can be naturally clamped at a limit") or setting limit on the prediction results? my first impression is the later but wanted to be sure. thanks.

**Comment by ngupta23** [comment]
Well, it was originally intended to be on the predictions only. The idea was that the input being fed would already be limited to the correct values. But the model knows nothing about these limits so it can produce forecasts (and prediction intervals) outside these limits. We would want to clamp those predictions and intervals to those limits.

One example is shown in the link in the 1st post. It talks about energy expended, but predictions show negative lower interval.

Do you see a use case for limiting the input data itself?

**Comment by vvvijaya** [comment]
Not really. As mentioned before, I just wanted to be sure the cap is for the predictions & intervals only. :) 

Thanks for clarifying!

**Comment by AJarman** [comment]
@ngupta23 happy to pick this up next after Croston, I have 2 weeks vacation from work from tomorrow.
Interested in some of the other issues here, but building up my confidence/knowledge of the library before implementing something more difficult!

**Comment by AJarman** [comment]
@ngupta23 I assume this would be useful in the regression module also?

**Comment by ngupta23** [comment]
Awesome. Please give it a shot. This is what I was thinking.

(1) Introduce 2 new parameters in `setup` called `lower_clamp` and `upper_clamp` (defaults to `None` which implies no limits). Note that this needs to be done in 2 files - `pycaret/time_series.py` (which supports the functional API) and `pycaret/internal/pycaret_experiment/time_series_experiment.py` (which supports the OOP API).
(2) Then in `pycaret/internal/pycaret_experiment/time_series_experiment.py` --> `predict_model`, you will need to clamp the "preds", "upper" and "lower" to these limits as appropriate.

**Comment by ngupta23** [comment]
> @ngupta23 I assume this would be useful in the regression module also?

I think it would be. @moezali1 @Yard1 Would you like to include this functionality in the regression module as well?

**Comment by AJarman** [comment]
> 

@ngupta23 Thanks for the heads up, very useful.
I'll have a look at the code this weekend , should be fairly straightforward - which branch should i use when forking ?
I assume `time-series` ?

**Comment by ngupta23** [comment]
Yes, fork from `time_series` and merge back to the same branch. Thanks!

**Comment by AJarman** [comment]
About to write some tests
I have just drafted this PR whilst I have time, please do not merge !

**Comment by ngupta23** [comment]
From @ltsaprounis  - https://robjhyndman.com/hyndsight/forecasting-within-limits/

Will add this transformer to sktime...

**Comment by ngupta23** [comment]
PR in sktime added: https://github.com/alan-turing-institute/sktime/pull/1913

**Comment by AJarman** [comment]
Happy to pick this up and integrate the`ScaledLogitTransformer` from sktime which @ltsaprounis kindly provided.
I would assume that this is now to be implemented differently during `setup` so the pipeline can persist(save) the transformer between `TimeSeriesExperiment`s


https://github.com/alan-turing-institute/sktime/releases/tag/v0.10.0
(now on pypi etc)
Assuming all the existing tests pass, is there anything else we need to consider when adding this newer version?

**Comment by ngupta23** [comment]
Hi @AJarman , Yes that is correct. It has to be added as a preprocessing step in the pipeline which will appropriately transform the data to keep it within the user specified limits.

Currently, preprocessing has not been implemented in the time-series module. We are actively working on it though (along with a revamp of the preprocessing for other modules such as regression/classification). 

https://github.com/pycaret/pycaret/pull/1921

Once that is done, we should be able to include `ScaledLogitTransformer` into the time series pipeline. So we have to wait until this PR is merged (we are about a month away from that - best guess at this time).

In the mean time, if you would like to contribute to any other issues, I will be happy to link you to a few options and provide guidance accordingly. Please let me know.

Thanks again for your patience.

**Comment by AJarman** [comment]
No problem, I will pick up something else !
I'll keep an eye on the issue above and implement once preprocessing is done.

**Comment by ngupta23** [comment]
Thank you. I appreicate your patience with this one.

**Comment by ngupta23** [comment]
@AJarman @ltsaprounis - Here is another option to limit the values but this uses clipping vs. transformation. I wonder which one is better?

https://github.com/alan-turing-institute/sktime/issues/2291#issuecomment-1077868581

**Comment by atsangarides** [comment]
Hi @ngupta23 just to clarify that I have that step to only make sure non-negative values are not fed into models that cannot handle that. I personally prefer having the "full" predictions and manually clipping these later according to the domain. Reason is unclipped predictions are more symmetric and represent model uncertainty better, whereas clipping can "fool" downstream users if that makes sense.
Anw commenting here as curious to follow the implementation.

**Comment by AJarman** [comment]
@ngupta23 I feel that both implementations have their use cases. Might be useful to have an upper limit, lower limit and 'clamp method' parameters. If the user simply wants to clamp predictions they could do this themselves but the implementation suggests we want to fit/evaluate the forecasters based on user-defined clamps

**Comment by ngupta23** [comment]
> just to clarify that I have that step to only make sure non-negative values are not fed into models that cannot handle that. I personally prefer having the "full" predictions and manually clipping these later according to the domain. 

@atsangarides @AJarman Thinking about this more, the use case in this issue is actually slightly different form what @atsangarides describes. Here we are saying that the data (naturally) does not exceed the limits. For example, energy consumption is always positive (> 0). So the training data will always be positive. But if the trend is decreasing, future predictions (or intervals) may be negative. The use case in this issue is just limiting the future predictions to be clamped to 0 for example by providing a lower clamp level.  

The use case that @atsangarides presents has negative values in the data itself. So it is a different use case. I don't think that is what we intend to support here.

> Reason is unclipped predictions are more symmetric and represent model uncertainty better, whereas clipping can "fool" downstream users if that makes sense.

This is true for the use case that you describe. For the one that is being addressed here, clamping probaby makes sense since metrics will get computed based on clamped values (0 in the example above) instead of negative values which is probably more representative.

Thoughts are welcome.

**Comment by ngupta23** [comment]
> If the user simply wants to clamp predictions they could do this themselves but the implementation suggests we want to fit/evaluate the forecasters based on user-defined clamps

@AJarman By the way, the way @ltsaprounis has implemented the transformer, if the training data itself is outside the limits specified by the user, the transformation will fail since the premise is that the training data is within the limits naturally and we are just trying to clamp the predictions to be within the same limits. @ltsaprounis - correct me if I am wrong.

**Comment by atsangarides** [comment]
I am actually dealing with energy demand/price forecasting :) 
I started clamping my predictions when they were actually made, but moving away from that, and handling clamping differently in different places.
If you use forecasts as inputs to an optimiser, you can use lower/upper quantiles for including uncertainty. Even though you cannot have <0 values for say demand, clipping your predictions could impact how you penalise different markets. All I am saying is that depending on what a user wants to use the forecasts for, clipping at an earlier stage might not be desirable. No  right or wrong answer 😄

**Comment by ngupta23** [comment]
@atsangarides - Got it. In any case, the way this will be implemented, user will be in control. They can decide to clamp the predictions or just leave it as is (default). So this would address both cases :)

**Comment by ngupta23** [comment]
@atsangarides Would love for you to try out PyCaret Time Series (based on sktime) and provide feedback. If you would be open to it, please connect on LinkedIn and I will follow up with you. Thanks!

https://www.linkedin.com/in/guptanick/

**Comment by atsangarides** [comment]
@ngupta23 we are the point of deciding on picking a framework Vs writing our own, so good timing. Your connections are maxed out so maybe reach out to [me](https://www.linkedin.com/in/andreas-tsangarides-2975a550/)

**Comment by AJarman** [comment]
Just summarising the implementation I did today:
## `oop.py, TSForecastingExperiment.setup()`
- Implemented two (optional) arguments to feed to `ScaledLogitTransformer`:
   - I have currently implemented this as `transform_target_kwargs` and `transform_exogenous_kwargs` (both dicts)
   - These set instance variables to `TSForecasting experiment` (implemented in `TSForecastingPreprocessor` parent class/mixin)
 ## `preprocess.py,  TSForecastingPreprocessor._add_transformation_steps`
- Added the `ScaledLogitTransformer` to `preprocess.py` within the transformers in `transform_dict`
- passed the kwargs to transformer constructors `TSForecastingPreprocessor._add_transformation_steps` from instance attributes. (Not sure whether to handle any possible errors here?)
  - I am happy to unpack the transform kwargs into upper/lower bounds in this implementation if there 
- I also removed the parentheses(calling `__init__`) from `transform_dict` factory in order to allow the kwargs to be passed later in the creation (other transformers don't appear to take any arguments)

(Note that this (like the other transformers, also can be used to transform exogenous variables.)

Still need to write some tests, feedback welcomed. 😄

**Comment by AJarman** [comment]
Bonus:
 `ExponentTransformer` and `SqrtTransformer` also accept kwargs, so this can help deliver that functionality as well.

**Comment by ngupta23** [comment]
@AJarman Could you submit a PR so we can review it (even one without tests would be ok for now until you add them later)? Thanks!

**Comment by AJarman** [comment]
> @AJarman Could you submit a PR so we can review it (even one without tests would be ok for now until you add them later)? Thanks!

I think the implementation is OK, will need a little while tomorrow evening to do tests for each of the transformers, which I'll probably work through in a notebook. It delivers a little more than the original issue 😀 !

**Comment by ngupta23** [comment]
Hi @AJarman 

I discussed this with @moezali1 and @Yard1 in the meeting on 20220409 and this is the recommendation. We should add an argument to setup called `forecast_limit: Optional[List[Union[int, float]] = None`. By default, no limits are applied (same as today's functionality). If user wants to set limits, they can do it as follows:

```python
forecast_limit = None # default - no limits
forecast_limit = [0, 10000000] # lower and upper limit
forecast_limit = [0, None]  # lower limit only
forecast_limit = [None, 10000000] # upper limit only
```

## Coding Actions:
- Inside `setup`, there will have to be a check to make sure that if a list is passed, it has exactly 2 entries and the types are correct.
- If this is enabled, we should add a step to the transformation pipeline. The order of transformations will be as follows
    - Impute
    - Limit
    - Transform
    - Scale  

## Future enhancements and user customization

In the future, we will also extend the following arguments to accept Transformer objects directly so any user wanting to set custom arguments, can pass the object directly (this is similar to what the other modules such as regression and classification do in pycaret). This will do away with the need for the kwargs in setup.

```python
transform_target = "box_cox"  # Today
transform_target = BoxCoxTransformer(limits = Sth, method="pearsonr")  # In the future (in lieu of transform_kwargs)

forecast_limit = [0, None] # current proposal
forecast_limit = ScaledLogitTransformer(0, None) # In the future (in lieu of transform_kwargs)
```

Let me know if this makes sense. Would be happy to discuss more if needed.

Thanks!

**Comment by AJarman** [comment]
That looks like a tidier implementation and probably easier to test !
Sorry for the slow reply I am on holiday and not really checking emails.
Just to confirm how the limits are applied- do we still want to use the `ScaledLogitTransformer` or just clip forecasts /predictions by implementing a custom pipeline member /metaestimator ?

**Comment by ngupta23** [comment]
> Sorry for the slow reply I am on holiday and not really checking emails.

No problem. Take your time.



> Just to confirm how the limits are applied- do we still want to use the `ScaledLogitTransformer` or just clip forecasts /predictions by implementing a custom pipeline member /metaestimator ?

This should be implemented using the `ScaledLogitTransformer` which should be inserted into the pipeline after the Imputer and before the Transformations such as Box-Cox and Scaling. Refer

> If this is enabled, we should add a step to the transformation pipeline. The order of transformations will be as follows
Impute
Limit <-- `ScaledLogitTransformer`
Transform
Scale

**Comment by AJarman** [comment]
`/workspaces/pycaret/.venv/lib/python3.8/site-packages/sktime/transformations/series/scaledlogit.py:129: RuntimeWarning: X in ScaledLogitTransformer should not have values greater than upper_bound`

I think I have encountered a problem and `ScaledLogitTransformer` may not be able to help- it seems that when you pass training values (in this case Y) above `upper_bound` or below `lower_bound` then it simply ignores those values (and converts them to `np.NaN`), whilst raising the warning above. (Using the airline dataset)

In many cases it also transforms to negative values, which I'm not sure many models will support. I encountered this whilst writing my first test, but I have documented this behaviour in [notebook:](https://github.com/pycaret/pycaret/blob/e78699bfcda651c6be0e72ec32be8b186deec7a3/time_series_1766_demo.ipynb)


![image](https://user-images.githubusercontent.com/26150780/163581738-06972abd-f04e-4235-a1d2-0350a6ccdb7c.png)

**Comment by ngupta23** [comment]
Hi @AJarman Thanks for the detailed eye on this one. This is what we can do

1. The intent for these limits is that the data is naturally limited to within those values and hence we want the forecasts to be limited within those values as well. Hence, we can assume that the input data is limited to be within those values. You can add a check in `setup` for this when the user enters limits and raise a `ValueError` (with an appropriate message) if the input data itself is outside these limits.
2. Regarding the negative values, I think that is fine. We take care of that appropriately internally by disabling certain model arguments

![image](https://user-images.githubusercontent.com/33585645/163849946-35889d1b-9390-4feb-b09c-ff8349b50564.png)

**Comment by moezali1** [comment]
Closing as part of the 4.0 backlog cleanup — see #4177.

PyCaret 4.0 is a ground-up rewrite (OOP-only API, sklearn-native engine, Python 3.11+, sklearn ≥ 1.7) and almost all pre-4.0 issues either reference removed APIs (`setup()`, `plot_model`, `evaluate_model`, `interpret_model`, `automl`, the legacy time-series module, the Streamlit dashboard, …) or no longer reproduce on the new code path.

**If this still affects you in 4.0, please re-file as a new issue** with:

- the experiment constructor call (`Experiment(...)`),
- the operation that fails,
- the full traceback,
- your Python / sklearn / pycaret versions.

Resources for 4.0:

- Docs: https://pycaret.org/docs
- Migrate from 3.x: https://pycaret.org/docs/guides/migrating-from-3-x
- API reference: https://pycaret.org/reference
- FAQ: https://pycaret.org/docs/resources/faqs

Thanks for filing — and apologies for the long latency. The repo had drifted; we're starting clean.
